gradle-plugin/src/main/groovy/org/groovymc/modsdotgroovy/gradle/MDGExtension.groovy

```diff
@@ -1,136 +1,377 @@
 package org.groovymc.modsdotgroovy.gradle
 
 import groovy.transform.CompileStatic
-import groovy.transform.PackageScope
+import org.gradle.api.Action
+import org.gradle.api.NamedDomainObjectProvider
 import org.gradle.api.Project
-import org.gradle.api.model.ObjectFactory
-import org.gradle.api.provider.MapProperty
+import org.gradle.api.artifacts.ConfigurablePublishArtifact
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.result.ResolvedArtifactResult
+import org.gradle.api.attributes.Bundling
+import org.gradle.api.attributes.Category
+import org.gradle.api.attributes.LibraryElements
+import org.gradle.api.attributes.Usage
+import org.gradle.api.attributes.java.TargetJvmEnvironment
+import org.gradle.api.file.CopySpec
+import org.gradle.api.file.FileCollection
+import org.gradle.api.provider.ListProperty
 import org.gradle.api.provider.Property
-import org.gradle.api.provider.SetProperty
+import org.gradle.api.provider.Provider
+import org.gradle.api.tasks.SourceSet
+import org.gradle.api.tasks.TaskProvider
+import org.gradle.language.jvm.tasks.ProcessResources
 import org.groovymc.modsdotgroovy.core.Platform
+import org.groovymc.modsdotgroovy.gradle.tasks.*
 
 import javax.inject.Inject
+import java.util.stream.Collectors
 
 @CompileStatic
-class MDGExtension {
-    @PackageScope static final String NAME = 'modsDotGroovy'
+abstract class MDGExtension {
+    private static final String EXPOSE_SOURCE_SET = 'shareModsDotGroovy'
+    private static final String DEFAULT_MDG = "mods.groovy"
 
-    private final MapProperty<Platform, String[]> platforms
-    private final SetProperty<String> environmentBlacklist
-    private final Property<Boolean> setupDsl
-    private final Property<Boolean> setupPlugins
-    private final Property<Boolean> setupTasks
-    //private final transient Project project
+    private static final String CONFIGURATION_NAME_ROOT = 'mdgRuntime'
+    private static final String CONFIGURATION_NAME_PLUGIN = 'mdgPlugin'
+    private static final String CONFIGURATION_NAME_FRONTEND = 'mdgFrontend'
+
+    private static final String MDG_MAVEN_GROUP = 'org.groovymc.modsdotgroovy'
+    private static final String MDG_FRONTEND_GROUP = MDG_MAVEN_GROUP + '.frontend-dsl'
+    private static final String MDG_PLUGIN_GROUP = MDG_MAVEN_GROUP + '.stock-plugins'
+
+    final Property<Boolean> setupDsl
+    final Property<Boolean> setupPlugins
+    final Property<Boolean> setupTasks
+    final Property<Boolean> inferGather
+    final ListProperty<Platform> platforms
+    final MDGConversionOptions conversionOptions
+    final Property<FileCollection> modsDotGroovyFile
+    final Multiplatform multiplatform
+
+    private final Property<Boolean> multiplatformFlag
+    private final SourceSet sourceSet
+
+    private final Project project
```

Comment:
I think storing the Project object without marking it as transient prevents the extension from being serializable, causing a compat issue with Gradle's configuration cache.

Reply:
Have you tested this? Extensions not used in task inputs don't get serialized

Reply:
Just tested it now - NeoGradle seems to force invalidating the configuration cache every run anyway at the moment so can't confirm, however I do recall NeoForged's GradleUtils changing their private Project field to transient to fix a configuration cache issue.

Reply:
You can test with the neoforge subproject that isn't NeoGradle. That one doesn't have any Minecraft-related plugin and thus should fully support configuration caching

Reply:
Your PR doesn't have a neoforge without neogradle subproject

Reply:
Oh lovely, they're on my DSL branch, right

Reply:
Can confirm, it works fine with configuration cache - this makes sense just given that the configuration cache will only serialize providers that are eventually stored as task inputs - and this extension never is, it just stores such providers.